kernel/src/vm/memfd.rs

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Memfd Implementation.
+
+use alloc::format;
+use core::sync::atomic::{AtomicU32, Ordering};
+
+use inherit_methods_macro::inherit_methods;
+
+use crate::{
+    events::IoEvents,
+    fs::{
+        file_handle::FileLike,
+        ramfs::new_detached_inode,
+        utils::{
+            AccessMode, FallocMode, Inode, InodeMode, IoctlCmd, Metadata, SeekFrom, StatusFlags,
+        },
+    },
+    prelude::*,
+    process::{
+        signal::{PollHandle, Pollable},
+        Gid, Uid,
+    },
+};
+
+/// Maximum file name length for `memfd_create`, excluding the final `\0` byte.
+///
+/// See <https://man7.org/linux/man-pages/man2/memfd_create.2.html>
+pub const MAX_MEMFD_NAME_LEN: usize = 249;
+
+pub struct MemfdFile {
+    inode: Arc<dyn Inode>,
+    #[expect(dead_code)]
+    name: String,
+    offset: Mutex<usize>,
+    access_mode: AccessMode,
+    status_flags: AtomicU32,
+}
+
+impl MemfdFile {
+    pub fn new(name: &str) -> Result<Self> {
+        if name.len() > MAX_MEMFD_NAME_LEN {
+            return_errno_with_message!(Errno::EINVAL, "MemfdManager: `name` is too long.");
+        }
+
+        let name = format!("/memfd:{} (deleted)", name);
+        let inode = new_detached_inode(
+            InodeMode::from_bits_truncate(0o777),
+            Uid::new_root(),
+            Gid::new_root(),
+        );
+
+        Ok(Self {
+            inode,
+            name,
+            offset: Mutex::new(0),
+            access_mode: AccessMode::O_RDWR,
+            status_flags: AtomicU32::new(0),
+        })
+    }
+}
+
+impl Pollable for MemfdFile {
+    fn poll(&self, mask: IoEvents, _poller: Option<&mut PollHandle>) -> IoEvents {
+        (IoEvents::IN | IoEvents::OUT) & mask
+    }
+}
+
+// TODO: Reuse the code from [`crate::fs::inode_handle::InodeHandle`].
+#[inherit_methods(from = "self.inode")]
+impl FileLike for MemfdFile {
+    fn read_at(&self, offset: usize, writer: &mut VmWriter) -> Result<usize>;
+    fn ioctl(&self, cmd: IoctlCmd, arg: usize) -> Result<i32>;
+    fn resize(&self, new_size: usize) -> Result<()>;
```

Comment:
The `resize` method needs be protected with the lock of `self.offset`. Otherwise, the `read` and `write` methods (and their variants) may use the invalidated values of `self.offset`.